File: modules/reduction/doc/hmsb.py
[ ## this file was manually modified by jt
    {
     'functor' : {
         'arity' : '1',
         'call_types' : [],
         'ret_arity' : '0',
         'rturn' : {
             'default' : 'T',
            },
         'special' : ['reduction'],  
         'type_defs' : [],
         'types' : ['real_', 'signed_int_', 'unsigned_int_']
        },
     'info' : 'manually modified',
     'unit' : {
         'global_header' : {
             'first_stamp' : 'created  by jt the 24/02/2011',
             'included' : [],   
             'simd_included' : ['#include <nt2/include/functions/bits.hpp>',
                                '#include <nt2/include/functions/shri.hpp>'],
             'cover_included' : ['#include <nt2/include/functions/bits.hpp>'],
             'no_ulp' : 'True',
             'notes' : [],
             'stamp' : 'modified by jt the 24/02/2011',
            },
         'ranges' : {
             'default' : [['nt2::Valmin<T>()', 'nt2::Valmax<T>()']],
            },
         'specific_values' : {
             'signed_int_' : {
                 'nt2::One<T>()' : {'result' : 'nt2::Zero<r_t>()','ulp_thresh' : '0',},
                 'nt2::Zero<T>()' : {'result' : 'nt2::Zero<r_t>()','ulp_thresh' : '0',},
                 'nt2::Signmask<T>()' : {'result' : 'r_t((1 << nt2::meta::cardinal_of<vT>::value) - 1)','ulp_thresh' : '0',},
                 'nt2::Allbits<T>()' : {'result' : 'r_t((1 << nt2::meta::cardinal_of<vT>::value) - 1)','ulp_thresh' : '0',},
                },
             'unsigned_int_' : {
                 'nt2::One<T>()' : {'result' : 'nt2::Zero<r_t>()','ulp_thresh' : '0',},
                 'nt2::Zero<T>()' : {'result' : 'nt2::Zero<r_t>()','ulp_thresh' : '0',},
                 'nt2::Allbits<T>()' : {'result' : 'r_t((1 << nt2::meta::cardinal_of<vT>::value) - 1)','ulp_thresh' : '0',},
                },
             'real_' : {
                 'nt2::Inf<T>()' : {'result' : 'nt2::Zero<r_t>()','ulp_thresh' : '0',},
                 'nt2::Minf<T>()' : {'result' : 'nt2::shri(nt2::Mone<nt2::int32_t>(),int(32-nt2::meta::cardinal_of<vT>::value))','ulp_thresh' : '0',},
                 'nt2::Mone<T>()' : {'result' : 'nt2::shri(nt2::Mone<nt2::int32_t>(),int(32-nt2::meta::cardinal_of<vT>::value))','ulp_thresh' : '0',},
                 'nt2::Nan<T>()' : {'result' : 'nt2::shri(nt2::Mone<nt2::int32_t>(),int(32-nt2::meta::cardinal_of<vT>::value))','ulp_thresh' : '0',},
                 'nt2::One<T>()' : {'result' : 'nt2::Zero<r_t>()','ulp_thresh' : '0',},
                 'nt2::Zero<T>()' : {'result' : 'nt2::Zero<r_t>()','ulp_thresh' : '0',},
                 'nt2::Signmask<T>()' : {'result' : 'r_t((1 << nt2::meta::cardinal_of<vT>::value) - 1)','ulp_thresh' : '0',},
                 'nt2::Allbits<T>()' : {'result' : 'r_t((1 << nt2::meta::cardinal_of<vT>::value) - 1)','ulp_thresh' : '0',},
                },
            },
         'verif_test' : {
             'nb_rand' : {
                 'default' : 'NT2_NB_RANDOM_TEST',
                },
             'property_call' : {
                 'default' : ['nt2::hmsb(a0)'],
                },
             'property_value' : {
                 'default' : ['nt2::b_and(a0, nt2::Signmask<r_t>()) != 0'],
                },
             'ulp_thresh' : {
                 'default' : ['0'],
                },
             'scalar_simul' :{
                    'default' : [
                        "        nt2::int32_t z = 0;",
                        "        nt2::uint32_t N = cardinal_of<n_t>::value;",
                        "        for(nt2::uint32_t i = 0; i< N; ++i)",
                        "        {",
                        "          z |= nt2::bits(a0[i]) >> (sizeof(iT)*CHAR_BIT - 1) << i;"
                        "        }",
                        "        NT2_TEST_EQUAL( v,z);",
                            ]
               },
            },
        },
    },
]

File: modules/boost/simd/reduction/unit/simd/hmsb.cpp
//==============================================================================
//         Copyright 2003 - 2012   LASMEA UMR 6602 CNRS/Univ. Clermont II
//         Copyright 2009 - 2012   LRI    UMR 8623 CNRS/Univ Paris Sud XI
//
//          Distributed under the Boost Software License, Version 1.0.
//                 See accompanying file LICENSE.txt or copy at
//                     http://www.boost.org/LICENSE_1_0.txt
//==============================================================================
#define NT2_UNIT_MODULE "nt2 boost.simd.reduction toolbox - hmsb/simd Mode"

//////////////////////////////////////////////////////////////////////////////
// unit test behavior of boost.simd.reduction components in simd mode
//////////////////////////////////////////////////////////////////////////////
/// created  by jt the 24/02/2011
///
#include <boost/simd/toolbox/reduction/include/functions/hmsb.hpp>
#include <boost/simd/sdk/simd/native.hpp>
#include <boost/simd/include/functions/bits.hpp>
#include <boost/simd/include/functions/shri.hpp>

#include <boost/dispatch/functor/meta/call.hpp>
#include <nt2/sdk/unit/tests/relation.hpp>
#include <nt2/sdk/unit/module.hpp>
#include <boost/simd/toolbox/constant/constant.hpp>

NT2_TEST_CASE_TPL ( hmsb_real__1_0,  BOOST_SIMD_SIMD_REAL_TYPES)
{
  using boost::simd::hmsb;
  using boost::simd::tag::hmsb_;
  using boost::simd::native;
  using boost::simd::meta::cardinal_of;
  typedef BOOST_SIMD_DEFAULT_EXTENSION  ext_t;
  typedef native<T,ext_t>                        vT;
  typedef typename boost::dispatch::meta::call<hmsb_(vT)>::type r_t;

  // specific values tests
  NT2_TEST_EQUAL(hmsb(boost::simd::Allbits<vT>()), r_t((1ull << boost::simd::meta::cardinal_of<vT>::value) - 1));
  NT2_TEST_EQUAL(hmsb(boost::simd::Inf<vT>()), boost::simd::Zero<r_t>());
  NT2_TEST_EQUAL(hmsb(boost::simd::Minf<vT>()), r_t(boost::simd::shri(boost::simd::Mone<boost::simd::int32_t>(),32-boost::simd::meta::cardinal_of<vT>::value)));
  NT2_TEST_EQUAL(hmsb(boost::simd::Mone<vT>()), r_t(boost::simd::shri(boost::simd::Mone<boost::simd::int32_t>(),32-boost::simd::meta::cardinal_of<vT>::value)));
  NT2_TEST_EQUAL(hmsb(boost::simd::Nan<vT>()), r_t(boost::simd::shri(boost::simd::Mone<boost::simd::int32_t>(),32-boost::simd::meta::cardinal_of<vT>::value)));
  NT2_TEST_EQUAL(hmsb(boost::simd::One<vT>()), boost::simd::Zero<r_t>());
  NT2_TEST_EQUAL(hmsb(boost::simd::Signmask<vT>()), r_t((1ull << boost::simd::meta::cardinal_of<vT>::value) - 1));
  NT2_TEST_EQUAL(hmsb(boost::simd::Zero<vT>()), boost::simd::Zero<r_t>());
} // end of test for real_

NT2_TEST_CASE_TPL ( hmsb_signed_int__1_0,  BOOST_SIMD_SIMD_INTEGRAL_SIGNED_TYPES)
{
  using boost::simd::hmsb;
  using boost::simd::tag::hmsb_;
  using boost::simd::native;
  using boost::simd::meta::cardinal_of;
  typedef BOOST_SIMD_DEFAULT_EXTENSION  ext_t;
  typedef native<T,ext_t>                        vT;
  typedef typename boost::dispatch::meta::call<hmsb_(vT)>::type r_t;

  // specific values tests
  NT2_TEST_EQUAL(hmsb(boost::simd::Allbits<vT>()), r_t((1ull << boost::simd::meta::cardinal_of<vT>::value) - 1));
  NT2_TEST_EQUAL(hmsb(boost::simd::One<vT>()), boost::simd::Zero<r_t>());
  NT2_TEST_EQUAL(hmsb(boost::simd::Signmask<vT>()), r_t((1ull << boost::simd::meta::cardinal_of<vT>::value) - 1));
  NT2_TEST_EQUAL(hmsb(boost::simd::Zero<vT>()), boost::simd::Zero<r_t>());
} // end of test for signed_int_

NT2_TEST_CASE_TPL ( hmsb_unsigned_int__1_0,  BOOST_SIMD_SIMD_UNSIGNED_TYPES)
{
  using boost::simd::hmsb;
  using boost::simd::tag::hmsb_;
  using boost::simd::native;
  using boost::simd::meta::cardinal_of;
  typedef BOOST_SIMD_DEFAULT_EXTENSION  ext_t;
  typedef native<T,ext_t>                        vT;
  typedef typename boost::dispatch::meta::call<hmsb_(vT)>::type r_t;

  // specific values tests
  NT2_TEST_EQUAL(hmsb(boost::simd::Allbits<vT>()), r_t((1ull << boost::simd::meta::cardinal_of<vT>::value) - 1));
  NT2_TEST_EQUAL(hmsb(boost::simd::One<vT>()), boost::simd::Zero<r_t>());
  NT2_TEST_EQUAL(hmsb(boost::simd::Zero<vT>()), boost::simd::Zero<r_t>());
} // end of test for unsigned_int_

NT2_TEST_CASE_TPL ( hmsb_logical, BOOST_SIMD_SIMD_TYPES)
{
  using boost::simd::hmsb;
  using boost::simd::tag::hmsb_;
  using boost::simd::native;
  using boost::simd::logical;
  using boost::simd::meta::cardinal_of;
  typedef BOOST_SIMD_DEFAULT_EXTENSION  ext_t;
  typedef native<logical<T>,ext_t>                          vT;
  typedef typename boost::dispatch::meta::call<hmsb_(vT)>::type r_t;

  // specific values tests
  NT2_TEST_EQUAL(hmsb(boost::simd::True<vT>())  , r_t((1ull << boost::simd::meta::cardinal_of<vT>::value) - 1));
  NT2_TEST_EQUAL(hmsb(boost::simd::False<vT>()) , boost::simd::Zero<r_t>());
}
